miden-crypto/src/merkle/smt/partial/mod.rs

```diff
@@ -0,0 +1,566 @@
+
+    /// Returns an opening of the leaf associated with `key`. Conceptually, an opening is a Merkle
+    /// path to the leaf, as well as the leaf itself.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if:
+    /// - the key is not tracked by this partial SMT.
+    pub fn open(&self, key: &Word) -> Result<SmtProof, MerkleError> {
+        let leaf = self.get_leaf(key)?;
+        let merkle_path = self.get_path(key);
+        Ok(SmtProof::new_unchecked(merkle_path, leaf))
+    }
+
+    /// Returns the leaf to which `key` maps.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if:
+    /// - the key is not tracked by this partial SMT.
+    pub fn get_leaf(&self, key: &Word) -> Result<SmtLeaf, MerkleError> {
+        self.get_tracked_leaf(key).ok_or(MerkleError::UntrackedKey(*key))
+    }
+
+    /// Returns the value associated with `key`.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if:
+    /// - the key is not tracked by this partial SMT.
+    pub fn get_value(&self, key: &Word) -> Result<Word, MerkleError> {
+        self.get_tracked_leaf(key)
+            .map(|leaf| leaf.get_value(key).unwrap_or_default())
+            .ok_or(MerkleError::UntrackedKey(*key))
+    }
+
+    // STATE MUTATORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Inserts a value at the specified key, returning the previous value associated with that key.
+    /// Recall that by definition, any key that hasn't been updated is associated with
+    /// [`Self::EMPTY_VALUE`].
+    ///
+    /// This also recomputes all hashes between the leaf (associated with the key) and the root,
+    /// updating the root itself.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if:
+    /// - the key is not tracked (see the type documentation for the definition of "tracked"). If an
+    ///   error is returned the tree is in the same state as before.
+    /// - inserting the key-value pair would exceed [`super::MAX_LEAF_ENTRIES`] (1024 entries) in
+    ///   the leaf.
+    pub fn insert(&mut self, key: Word, value: Word) -> Result<Word, MerkleError> {
+        let current_leaf = self.get_tracked_leaf(&key).ok_or(MerkleError::UntrackedKey(key))?;
+        let leaf_index = current_leaf.index();
+        let previous_value = current_leaf.get_value(&key).unwrap_or(EMPTY_WORD);
+        let prev_entries = current_leaf.num_entries();
+
+        let leaf = self
+            .leaves
+            .entry(leaf_index.value())
+            .or_insert_with(|| SmtLeaf::new_empty(leaf_index));
+
+        if value != EMPTY_WORD {
+            leaf.insert(key, value).map_err(|e| match e {
+                SmtLeafError::TooManyLeafEntries { actual } => {
+                    MerkleError::TooManyLeafEntries { actual }
+                },
+                other => panic!("unexpected SmtLeaf::insert error: {:?}", other),
+            })?;
+        } else {
+            leaf.remove(key);
```

Comment:
Should we not check here if the leaf became empty, and if so, remove the leaf from `self.leaves`? If so, let's make sure we have a test that checks this scenario.